// victory-core/lib/victory-primitives/bar.js
Object.defineProperty(exports,"__esModule",{value:true});var _assign2=require("lodash/assign");var _assign3=_interopRequireDefault(_assign2);var _extends=Object.assign||function(target){for(var i=1;i<arguments.length;i++){var source=arguments[i];for(var key in source){if(Object.prototype.hasOwnProperty.call(source,key)){target[key]=source[key];}}}return target;};var _createClass=function(){function defineProperties(target,props){for(var i=0;i<props.length;i++){var descriptor=props[i];descriptor.enumerable=descriptor.enumerable||false;descriptor.configurable=true;if("value"in descriptor)descriptor.writable=true;Object.defineProperty(target,descriptor.key,descriptor);}}return function(Constructor,protoProps,staticProps){if(protoProps)defineProperties(Constructor.prototype,protoProps);if(staticProps)defineProperties(Constructor,staticProps);return Constructor;};}();var _react=require("react");var _react2=_interopRequireDefault(_react);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _classCallCheck(instance,Constructor){if(!(instance instanceof Constructor)){throw new TypeError("Cannot call a class as a function");}}function _possibleConstructorReturn(self,call){if(!self){throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call&&(typeof call==="object"||typeof call==="function")?call:self;}function _inherits(subClass,superClass){if(typeof superClass!=="function"&&superClass!==null){throw new TypeError("Super expression must either be null or a function, not "+typeof superClass);}subClass.prototype=Object.create(superClass&&superClass.prototype,{constructor:{value:subClass,enumerable:false,writable:true,configurable:true}});if(superClass)Object.setPrototypeOf?Object.setPrototypeOf(subClass,superClass):subClass.__proto__=superClass;}var


Bar=function(_React$Component){_inherits(Bar,_React$Component);function Bar(){_classCallCheck(this,Bar);return _possibleConstructorReturn(this,(Bar.__proto__||Object.getPrototypeOf(Bar)).apply(this,arguments));}_createClass(Bar,[{key:"getVerticalBarPath",value:function getVerticalBarPath(





















props,width){var
x=props.x,y0=props.y0,y=props.y;
var size=width/2;
return"M "+(x-size)+", "+y0+"\n      L "+(
x-size)+", "+y+"\n      L "+(
x+size)+", "+y+"\n      L "+(
x+size)+", "+y0+"\n      L "+(
x-size)+", "+y0;
}},{key:"getHorizontalBarPath",value:function getHorizontalBarPath(

props,width){var
x=props.x,y0=props.y0,y=props.y;
var size=width/2;
return"M "+y0+", "+(x-size)+"\n      L "+
y0+", "+(x+size)+"\n      L "+
y+", "+(x+size)+"\n      L "+
y+", "+(x-size)+"\n      L "+
y0+", "+(x-size);
}},{key:"getBarPath",value:function getBarPath(

props,width){
return this.props.horizontal?
this.getHorizontalBarPath(props,width):this.getVerticalBarPath(props,width);
}},{key:"getBarWidth",value:function getBarWidth(

props){var
style=props.style,width=props.width,data=props.data;
var padding=props.padding.left||props.padding;
var defaultWidth=data.length===0?8:(width-2*padding)/data.length;
return style&&style.width?style.width:defaultWidth;
}},{key:"renderBar",value:function renderBar(


path,style,events){var _props=
this.props,role=_props.role,shapeRendering=_props.shapeRendering;
return(
_react2.default.createElement("path",_extends({
d:path,
style:style,
role:role||"presentation",
shapeRendering:shapeRendering||"auto"},
events)));


}},{key:"render",value:function render()

{

var barWidth=this.getBarWidth(this.props);
var path=typeof this.props.x==="number"?
this.getBarPath(this.props,barWidth):undefined;
var style=(0,_assign3.default)({fill:"black",stroke:"none"},this.props.style);
return this.renderBar(path,style,this.props.events);
}}]);return Bar;}(_react2.default.Component);Bar.propTypes={datum:_react.PropTypes.object,events:_react.PropTypes.object,horizontal:_react.PropTypes.bool,index:_react.PropTypes.number,role:_react.PropTypes.string,scale:_react.PropTypes.object,shapeRendering:_react.PropTypes.string,style:_react.PropTypes.object,x:_react.PropTypes.number,y:_react.PropTypes.number,y0:_react.PropTypes.number,width:_react.PropTypes.number,padding:_react.PropTypes.oneOfType([_react.PropTypes.number,_react.PropTypes.object]),data:_react.PropTypes.array};exports.default=Bar;
